File: leetCode/Medium/0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int m = 0;
        if(nums.length == 1) {return nums[0];}
        else if(nums.length == 2) return Math.max(nums[0], nums[1]);
        else {
            int bBefore = nums[0];
            int before = Math.max(nums[1],nums[0]);
            for(int i=2; i<nums.length; i++) {
                m = Math.max(before, bBefore+nums[i]);
                bBefore = before;
                before = m;
            }
        }
        return m;
    }
}

File: leetCode/Medium/0198-house-robber/0198-house-robber.kt
class Solution {
    fun rob(nums: IntArray): Int {
        if(nums.size == 1) return nums[0]
    
        val r = IntArray(nums.size){0}
        r[0] = nums[0]
        r[1] = Math.max(nums[0], nums[1])
        
        for(i in 2..nums.lastIndex) {
            r[i] = Math.max(r[i-2]+nums[i], r[i-1])
        }
        
        return r[nums.lastIndex]
    }
}